File: go-backend-interview-prep/04_interfaces_and_generics/09_generics_vs_interfaces/main.go
package main

import (
	"fmt"
	"strings"
)

// ============================================================
// APPROACH 1: INTERFACES (runtime polymorphism, behavior-based)
// ============================================================

// Formatter interface: different types format differently.
type Formatter interface {
	Format() string
}

type PlainText struct{ Content string }

func (p PlainText) Format() string { return p.Content }

type HTMLText struct{ Content string }

func (h HTMLText) Format() string {
	return "<p>" + h.Content + "</p>"
}

// Works with ANY Formatter -- behavior is the contract.
func PrintFormatted(f Formatter) {
	fmt.Println(" ", f.Format())
}

// ============================================================
// APPROACH 2: GENERICS (compile-time, type-based)
// ============================================================

// Reverse works on any slice type -- same logic, different types.
func Reverse[T any](s []T) []T {
	result := make([]T, len(s))
	for i, v := range s {
		result[len(s)-1-i] = v
	}
	return result
}

// Unique removes duplicates -- needs comparable constraint.
func Unique[T comparable](s []T) []T {
	seen := make(map[T]bool)
	var result []T
	for _, v := range s {
		if !seen[v] {
			seen[v] = true
			result = append(result, v)
		}
	}
	return result
}

// ============================================================
// WHEN INTERFACE IS BETTER
// ============================================================

// Service layer: behavior-based. Different implementations
// for production, testing, etc.
type UserStore interface {
	GetUser(id int) string
}

type DBStore struct{}

func (d DBStore) GetUser(id int) string {
	return fmt.Sprintf("User-%d (from DB)", id)
}

type MockStore struct{}

func (m MockStore) GetUser(id int) string {
	return fmt.Sprintf("MockUser-%d", id)
}

func greetUser(store UserStore, id int) {
	fmt.Println("  Hello,", store.GetUser(id))
}

func main() {
	// --- Interface approach: behavior polymorphism ---
	fmt.Println("=== INTERFACE: different behavior ===")
	PrintFormatted(PlainText{Content: "hello world"})
	PrintFormatted(HTMLText{Content: "hello world"})

	// --- Generic approach: same logic, different types ---
	fmt.Println("\n=== GENERIC: same logic, different types ===")
	fmt.Println("reversed ints:", Reverse([]int{1, 2, 3, 4, 5}))
	fmt.Println("reversed strs:", Reverse([]string{"a", "b", "c"}))

	fmt.Println("unique ints:", Unique([]int{1, 2, 2, 3, 3, 3}))
	fmt.Println("unique strs:", Unique([]string{"go", "go", "rust"}))

	// --- Interface for service abstraction ---
	fmt.Println("\n=== INTERFACE: service abstraction ===")
	greetUser(DBStore{}, 1)
	greetUser(MockStore{}, 1) // easy to swap for tests

	// --- Decision framework ---
	fmt.Println("\n=== WHEN TO USE WHICH ===")
	decisions := []string{
		"Interface: behavior differs between types (Formatter, Store)",
		"Generic:   logic is identical, types differ (Reverse, Unique)",
		"Interface: dependency injection, mocking, service layers",
		"Generic:   type-safe containers, algorithms, transformations",
		"Default:   start with interface, add generics only if needed",
	}
	for _, d := range decisions {
		fmt.Println(" ", d)
	}
	_ = strings.Builder{} // avoid unused import
}

File: go-backend-interview-prep/04_interfaces_and_generics/09_generics_vs_interfaces/main.py
# Python equivalent of Generics vs Interfaces (compare with main.go)
# In Python, duck typing and ABC/Protocol fill the "interface" role.
# typing module fills the "generics" role (but is not enforced at runtime).

from typing import TypeVar, Protocol

T = TypeVar("T")


# ============================================================
# APPROACH 1: PROTOCOL (structural typing, like Go interface)
# ============================================================

class Formatter(Protocol):
    def format(self) -> str: ...


class PlainText:
    def __init__(self, content: str):
        self.content = content

    def format(self) -> str:
        return self.content


class HTMLText:
    def __init__(self, content: str):
        self.content = content

    def format(self) -> str:
        return f"<p>{self.content}</p>"


def print_formatted(f: Formatter) -> None:
    print(" ", f.format())


# ============================================================
# APPROACH 2: GENERIC-STYLE (same logic, different types)
# ============================================================

def reverse(s: list[T]) -> list[T]:
    return s[::-1]


def unique(s: list) -> list:
    seen = set()
    result = []
    for v in s:
        if v not in seen:
            seen.add(v)
            result.append(v)
    return result


# ============================================================
# PROTOCOL FOR SERVICE ABSTRACTION
# ============================================================

class UserStore(Protocol):
    def get_user(self, id: int) -> str: ...


class DBStore:
    def get_user(self, id: int) -> str:
        return f"User-{id} (from DB)"


class MockStore:
    def get_user(self, id: int) -> str:
        return f"MockUser-{id}"


def greet_user(store: UserStore, id: int) -> None:
    print("  Hello,", store.get_user(id))


def main():
    # --- Protocol approach: behavior polymorphism ---
    print("=== PROTOCOL: different behavior ===")
    print_formatted(PlainText("hello world"))
    print_formatted(HTMLText("hello world"))

    # --- Generic-style: same logic, different types ---
    print("\n=== GENERIC: same logic, different types ===")
    print("reversed ints:", reverse([1, 2, 3, 4, 5]))
    print("reversed strs:", reverse(["a", "b", "c"]))

    print("unique ints:", unique([1, 2, 2, 3, 3, 3]))
    print("unique strs:", unique(["go", "go", "rust"]))

    # --- Protocol for service abstraction ---
    print("\n=== PROTOCOL: service abstraction ===")
    greet_user(DBStore(), 1)
    greet_user(MockStore(), 1)

    # --- Decision framework ---
    print("\n=== WHEN TO USE WHICH ===")
    decisions = [
        "Protocol: behavior differs (Formatter, Store)",
        "Generics: logic identical, types differ (reverse, unique)",
        "Protocol: dependency injection, mocking",
        "Generics: type-safe containers, algorithms",
        "Default:  Python duck typing handles most cases naturally",
    ]
    for d in decisions:
        print(" ", d)


if __name__ == "__main__":
    main()
